### DISCUSSION TITLE
CogAlg Review

### DISCUSSION THREAD

**Discussion by ASBJ021** [boris-kz/CogAlg#70] [General]
Hi Chris and Boris,

I would like to thank you for giving me this opportunity. I was curious when I learned from Chris that you are working on CogAlgo which will be an alternative to CNN and DNN, so I took some time to go through your github  and tried to understand the algorithm. 

This is my understanding from line_pattern and some questions:

1.  A binary or grayscale image will be input to the code. Also some pattern filters or hyper parameters are initialized. 
     For ex: ave = 15, ave_min = 2, ave_M = 50, etc.
     I have one question on what basis these parameters are given these value ? 

2.  Then the image is passed to cross_comp funcion, where we take frame of pixels(row wise , from what I understood) and 
     convert the to frame of patterns.

        - Cross_comp:  In this function frame of pixels (row wise) are converted to frame of patterns.  A binary or gray scale image is 
                                 input. 

                             - Frame of pixels is an array of each row of pixel value of the image.
                             - dert_  value is calculated, which is the tuple of  ‘p’, ‘d’,  & ‘m’. 
                                     Where , 
		                             ‘p’ is previous pixel value ,
		                             ‘m’ is ave-abs(d) ,
		                             ‘d’ is difference between adjacent pixel values ,
                            - Then, dert_ is passed to form_pm function , which returns Pm_ values.
                            - A list of Pm_  values forms an array of frames of pattern.
                            - If Length (Pm_) > 4:
                                             - Adj_M is calculated.
                                             - intra_Pm_ is called which evaluates for sub_recursions per Pm_.
              - form_Pm_ : It takes dert_ as input and output the Pm_  values based on ‘m’ signs.
                          - sign = dert.m>0 , 
                          - dert =  P_dert 
                          - Returns an array which in formed by pixels with same ‘m’  sign.

             - form_Pd_ :  output the Pd_  values based on ‘d’ sign.
                            - sign = dert.d>0 , 
                            - dert =  P_dert 
                            - Returns an array which in formed by pixels with same ‘d’  sign.

            - Cluster Class: it is used to define structure for CP and Cdert. (from what I understood).
            - I want to ask that why did you choose this method over other methods.

 3. These frame of patterns will be used to form the new frame of patterns until a result is obtained.

I wanted to discuss and ask more things.  I think we can use this algorithm for object detection, object classification. Use it for the ADAS system, or maybe for NLP and many more applications.

I tried to accumulate my thought and express them in a simple way as possible.  

Thanks,
Akash

**Comment by boris-kz** [comment]
Thanks for taking a look Akash!

Line_patterns is not supposed work on it's own, it's only the 1st level of an indefinitely deep hierarchy, and only in 1D.
Have gone through this: https://github.com/boris-kz/CogAlg/wiki?
If you want understand the principles, that would be in Readme, you are welcome to discuss it here.

There should be tons of applications, but the alg is far from complete, we can't even properly test it yet.

**Reply by ASBJ021** [reply]
Thanks for the response. I will look into the readme file and other code files to understand it more deeply.

**Reply by ASBJ021** [reply]
Hi Boris,
I have gone through the link you provided. My understanding of algorithm became more clear. Now I am curious and interested in this algorithm. Is their any way I can contribute in it? If any, then how ? Can you guide me ? ( I have read contributions file, but specifically I dont know where to start )

Thanks,
Akash.

**Comment by boris-kz** [comment]
Thanks Akash, we are currently working on line_PPs_draft in 1D, comp_slice_ and comp_blob_draft in 2D.
Have you forked my repo yet?

**Reply by ASBJ021** [reply]
Yes, I have forked this repo & currently working on it. Hope I would be able to work on something.
